File: Haskell/05-cheapest-permutation.hs
import Control.Monad (replicateM)
import Data.Array (Array, listArray, (!))
import Data.List (permutations)


convertToIntArray :: Int -> String -> Array Int Int
convertToIntArray n string = listArray (0, n - 1) $ map read $ words string


readMatrix :: Int -> IO (Array Int (Array Int Int))
readMatrix n = do
    lines' <- replicateM n getLine
    return $ listArray (0, n - 1) $ map (convertToIntArray n) lines'


findCheapestPermutation :: (Array Int (Array Int Int)) -> Int -> [Int]
findCheapestPermutation matrix n =
    let xs = [0..(n-1)]
        calcCost :: [Int] -> Int
        calcCost perm = foldl (\cost (i, j) -> cost + matrix ! i ! j) 0 $ zip perm (tail perm)
        minPerm = fst $ foldl (\acc @ (_, minCost) p ->
            let cost = calcCost p in if cost < minCost then (p, cost) else acc) (xs, calcCost xs) $ permutations xs
    in if n == 1 then xs else minPerm


main :: IO()
main = do
    n <- readLn
    matrix <- readMatrix n
    let result = findCheapestPermutation matrix n
    putStrLn $ unwords $ map (show . (+ 1)) result

File: Python/05_cheapest_permutation.py
import sys
import itertools as it


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def read_matrix(reader, n):
    matrix = []
    for _ in range(n):
        line = next(reader)
        row = convert_to_intlist(line)
        matrix.append(row)
    return matrix


def find_cheapest_permutation(matrix, n):
    def calc_cost(perm):
        cost = 0
        for i, j in zip(perm, perm[1:]):
            cost += matrix[i][j]
        return cost

    min_perm = list(range(n))
    min_cost = calc_cost(min_perm)
    perms = it.permutations(min_perm)
    for p in perms:
        cost = calc_cost(p)
        if cost < min_cost:
            min_cost = cost
            min_perm = p
    return min_perm


def main():
    reader = sys.stdin
    n = int(next(reader))
    matrix = read_matrix(reader, n)
    result = find_cheapest_permutation(matrix, n)
    print(' '.join([str(elem + 1) for elem in result]))


if __name__ == '__main__':
    main()
